ginivision/src/main/java/net/gini/android/vision/camera/CameraFragmentImpl.java

```diff
@@ -319,17 +333,120 @@ public void run() {
+                    }
+
+                    @Override
+                    public void permissionDenied() {
+                        LOG.info("Read storage permission denied");
+                        showStoragePermissionDeniedDialog();
+                    }
+
+                    @Override
+                    public void shouldShowRequestPermissionRationale(
+                            @NonNull final RationaleResponse response) {
+                        LOG.info("Show read storage permission rationale");
+                        showStoragePermissionRationale(response);
+                    }
+                });
             }
         });
     }
 
+    private void showStoragePermissionRationale(
+            @NonNull final PermissionRequestListener.RationaleResponse response) {
+        mFragment.showAlertDialog(R.string.gv_storage_permission_rationale,
+                R.string.gv_storage_permission_rationale_positive_button,
+                new DialogInterface.OnClickListener() {
+            @Override
+            public void onClick(final DialogInterface dialogInterface,
+                    final int i) {
+                LOG.info("Requesting storage permission from rationale");
+                response.requestPermission();
+            }
+        });
+    }
+
+    private void showStoragePermissionDeniedDialog() {
+        mFragment.showAlertDialog(R.string.gv_storage_permission_denied,
+                R.string.gv_storage_permission_denied_positive_button,
+                new DialogInterface.OnClickListener() {
+                    @Override
+                    public void onClick(
+                            final DialogInterface dialogInterface,
+                            final int i) {
+                        LOG.info("Open app details in Settings app");
+                        showAppDetailsSettingsScreen();
+                    }
+                }, R.string.gv_storage_permission_denied_negative_button);
+    }
+
+    private void showFileChooser() {
+        LOG.debug("Importing document");
+        final Activity activity = mFragment.getActivity();
+        if (activity == null) {
+            return;
+        }
+        Intent fileChooserIntent = FileChooserActivity.createIntent(activity);
+        mFragment.startActivityForResult(fileChooserIntent, REQ_CODE_CHOOSE_FILE);
+    }
+
+    private void showAppDetailsSettingsScreen() {
+        final Activity activity = mFragment.getActivity();
+        if (activity == null) {
+            return;
+        }
+        final Intent intent = new Intent(
+                Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
+        final Uri uri = Uri.fromParts("package",
+                activity.getPackageName(), null);
+        intent.setData(uri);
+        activity.startActivity(intent);
+    }
+
+    private void requestStoragePermission(@NonNull final PermissionRequestListener listener) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
```

Comment:
Check for M is done in RuntimePermissions and this I added because READ_EXTERNAL_STORAGE is available since Jelly Bean and you are totally right it makes no sense making the trip to RuntimePermissions and back on pre-M 😁